### PR TITLE
Support hooks in storyshots

### PR DESCRIPTION
**Pull request by brandon-leapyear**
:sparkles: _**This is an old work account. Please reference @brandonchinn178 for all future communication**_ :sparkles:

---

Issue: #10483 

## What I did
Rendered `storyFn` as a component in storyshots, just like when rendering it for storybook.

### CODE REVIEW

**Review by shilman — CHANGES_REQUESTED** [review]
Thanks @brandon-leapyear ! The update looks great, but it's actually modifying the output snapshots so that they're wrapped with `<storyFn>...</storyFn>`. If you can fix it to preserve the old output, that would be a lot better.

**Comment by brandon-leapyear** [comment]
unfortunately, that is one of the side effects of this fix; you can see the outputs also changed in the original PR into storybook.

my goal for this PR was to make storyshots have the same capability as storybook's normal usage. otherwise, id have something working in storybook, but breaking with storyshots.

i think if the output for this solution is desired, the discussion should be propagated back to the storybook package

File: addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts [hunk 1]
```diff
@@ -1,8 +1,10 @@
 // eslint-disable-next-line import/no-extraneous-dependencies
+import React from 'react';
```
Comment (shilman) [inline addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts:2]: Can you move this line above the previous one? It's causing this linting error:

```
/tmp/storybook/addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts
  3:1  error  'react-test-renderer' should be listed in the project's dependencies. Run 'npm i -S react-test-renderer' to add it  import/no-extraneous-dependencies
```

**Comment by shilman** [comment]
@brandon-leapyear I understand that. The issue is that this will break everybody's existing snapshots. It's not the end of the world because we're in a 6.0 major release right now, but it would be better if we could find a workaround. cc @Hypnosphi

**Comment by brandon-leapyear** [comment]
I see. That makes sense. I think that decision is higher level than what I can offer.

FWIW, I used this hook in my project, and I don't see `<storyFn>` in my snapshots

**Comment by ndelangen** [comment]
I think there's something to be done @shilman...

File: addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts [hunk 2]
```diff
@@ -1,8 +1,11 @@
...
+// eslint-disable-next-line import/no-extraneous-dependencies
 import reactTestRenderer from 'react-test-renderer';
 
 function getRenderedTree(story: any, context: any, { renderer, ...rendererOptions }: any) {
-  const storyElement = story.render();
+  const StoryFn = story.render;
+  const storyElement = React.createElement(StoryFn);
   const currentRenderer = renderer || reactTestRenderer.create;
   const tree = currentRenderer(storyElement, rendererOptions);
```
Comment (ndelangen) [inline addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts:11]: Instead of returning the full tree, we could return the first child here?

**Review by ndelangen — CHANGES_REQUESTED** [review]

File: addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts (hunk 2 above)
Comment (ndelangen) [reply addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts:11]: on line 12

Comment (shilman) [reply addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts:11]: 🧠

Comment (ndelangen) [reply addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts:11]: @brandon-leapyear if you want to do that that'd be amazing! if not, that's fine; I can take it from here?.

Comment (brandon-leapyear) [reply addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts:11]: I did a thing. Not sure if it's the correct change (is there an API we expect from `currentRenderer`?), so please let me know if it should be different

Comment (ndelangen) [reply addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts:11]: I'd assume it should be the first child, rather then all children?

Did you update the snapshots?

Comment (brandon-leapyear) [reply addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts:11]: Yeah, the snapshots are up to date. `.children()` doesn't seem like it returns an array, so I wasn't sure how to get the first child

Comment (ndelangen) [reply addons/storyshots/storyshots-core/src/frameworks/react/renderTree.ts:11]: Can i get access to push to your remote/branch?

I have a fix

**Comment by shilman** [comment]
Released https://github.com/storybookjs/storybook/releases/tag/v6.0.0-alpha.43 which references this issue in #10529 . Closing this!

**Pull request merged**